fs-cache/src/memory_limited_storage.rs

```diff
@@ -0,0 +1,273 @@
+        + std::hash::Hash
+        + std::str::FromStr,
+    V: Clone + serde::Serialize + serde::de::DeserializeOwned,
+{
+    pub fn new(
+        label: String,
+        path: &Path,
+        max_memory_bytes: usize,
+    ) -> Result<Self> {
+        let mut storage = Self {
+            label,
+            path: PathBuf::from(path),
+            memory_cache: LinkedHashMap::new(),
+            current_memory_bytes: 0,
+            max_memory_bytes,
+        };
+
+        storage.load_fs()?;
+
+        Ok(storage)
+    }
+
+    pub fn label(&self) -> String {
+        self.label.clone()
+    }
+
+    pub fn get(&mut self, key: &K) -> Option<V> {
+        // Check memory cache first - will update LRU order automatically
+        if let Some(value) = self.memory_cache.get_refresh(key) {
+            return Some(value.clone());
+        }
+
+        // Try to load from disk
+        let file_path = self.path.join(format!("{}.json", key));
+        if file_path.exists() {
+            // Doubt: Update file's modiied time (in disk) on read to preserve LRU across app restarts?
+            match self.load_value_from_disk(key) {
+                Ok(value) => {
+                    self.add_to_memory_cache(key.clone(), value.clone());
+                    Some(value)
+                }
+                Err(err) => {
+                    log::error!(
+                        "{} cache: failed to load key={}: {}",
+                        self.label,
+                        key,
+                        err
+                    );
+                    None
+                }
+            }
+        } else {
+            None
+        }
+    }
+
+    pub fn set(&mut self, key: K, value: V) -> Result<()> {
+        // Always write to disk first
+        self.write_value_to_disk(&key, &value)?;
+
+        // Then update memory cache
+        self.add_to_memory_cache(key, value);
+
+        Ok(())
+    }
+
+    pub fn load_fs(&mut self) -> Result<()> {
+        if !self.path.exists() {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "Folder does not exist".to_owned(),
+            ));
+        }
+
+        if !self.path.is_dir() {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "Path is not a directory".to_owned(),
+            ));
+        }
+
+        // First pass: collect metadata only
+        let mut file_metadata = Vec::new();
+        for entry in fs::read_dir(&self.path)? {
+            let entry = entry?;
+            let path = entry.path();
+            if path.is_file()
+                && path
+                    .extension()
+                    .map_or(false, |ext| ext == "json")
+            {
+                if let Ok(metadata) = fs::metadata(&path) {
+                    let key = extract_key_from_file_path(&self.label, &path)?;
+                    file_metadata.push((key, metadata.len() as usize));
+                }
```

Comment:
This assumes that every file with valid JSON is considered a valid key. Is this an intentional assumption?

Reply:
Yes, initially, I had made this assumption. However, since our main objective for this cache is to store large values like bitmaps, images, etc., it is no longer needed. Thank you for pointing it out.